File: javascript/src/local/context.ts
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";

import {
  classifyLocalPathSensitivity,
  type LocalFileDeliver,
  type LocalGrepResponse,
  type LocalPathSensitivity,
  type LocalPathSensitivityInfo,
  type LocalSummary,
  type LocalWorkspace,
  type LocalWorkspaceSnapshotFile,
} from "./core.js";

export interface LocalContextPackageParams {
  path?: string;
  include?: string[];
  exclude?: string[];
  query?: string;
  maxFiles?: number;
  maxBytes?: number;
  maxBytesPerFile?: number;
  previewBytes?: number;
  includeContent?: boolean;
  includeHashes?: boolean;
  includeSummary?: boolean;
  includeSearch?: boolean;
  includeSecrets?: boolean;
}

export interface LocalContextFile {
  path: string;
  size: number;
  sha256?: string;
  mime_type?: string;
  sensitivity: LocalPathSensitivity;
  sensitivity_reason?: string;
  content?: string;
  content_base64?: string;
  encoding?: LocalFileDeliver["encoding"];
  truncated?: boolean;
  omitted_reason?: string;
}

export interface LocalContextManifest {
  object: "local_context_manifest";
  root: string;
  workspace_name: string;
  generated_at_unix: number;
  base_path: string;
  file_count: number;
  total_bytes: number;
  included_bytes: number;
  truncated: boolean;
  files: LocalContextFile[];
  summary?: LocalSummary;
  search?: LocalGrepResponse;
}

export async function createLocalContextPackage(
  workspace: LocalWorkspace,
  params: LocalContextPackageParams = {},
): Promise<LocalContextManifest> {
  const basePath = params.path ?? ".";
  const maxFiles = positiveInt(params.maxFiles, 80);
  const maxBytes = positiveInt(params.maxBytes, 256 * 1024);
  const maxBytesPerFile = positiveInt(params.maxBytesPerFile, 32 * 1024);
  const previewBytes = positiveInt(params.previewBytes, maxBytesPerFile);
  const includeContent = params.includeContent ?? true;
  const includeHashes = params.includeHashes ?? true;
  const includeSummary = params.includeSummary ?? true;
  const includeSearch = Boolean(params.includeSearch && params.query?.trim());
  const includeSecrets = params.includeSecrets ?? false;

  const stats = await workspace.list(basePath, { recursive: true });
  const fileStats = stats
    .filter((item) => item.type === "file")
    .filter((item) => matchesPathFilters(item.path, params.include, params.exclude))
    .sort((a, b) => a.path.localeCompare(b.path));

  const files: LocalContextFile[] = [];
  let includedBytes = 0;
  let truncated = fileStats.length > maxFiles;

  for (const item of fileStats.slice(0, maxFiles)) {
    const sensitivity = classifyLocalPathSensitivity(item.path);
    const baseFile: LocalContextFile = {
      path: item.path,
      size: item.size,
      sensitivity: sensitivity.sensitivity,
      sensitivity_reason: sensitivity.reason,
    };

    if (!includeSecrets && sensitivity.sensitivity === "secret") {
      files.push({ ...baseFile, omitted_reason: "secret_path" });
      continue;
    }
    if (item.size > maxBytesPerFile) {
      files.push({ ...baseFile, omitted_reason: "file_too_large" });
      truncated = true;
      continue;
    }
    if (includedBytes >= maxBytes) {
      files.push({ ...baseFile, omitted_reason: "package_budget_exceeded" });
      truncated = true;
      continue;
    }

    const remaining = maxBytes - includedBytes;
    const readBudget = Math.min(previewBytes, maxBytesPerFile, remaining);
    if (readBudget <= 0) {
      files.push({ ...baseFile, omitted_reason: "package_budget_exceeded" });
      truncated = true;
      continue;
    }

    const delivered = await workspace.readFile(item.path, { maxBytes: readBudget });
    includedBytes += delivered.encoding === "text"
      ? Buffer.byteLength(delivered.content ?? "", "utf8")
      : Buffer.byteLength(delivered.content_base64 ?? "", "base64");

    const packaged: LocalContextFile = {
      ...baseFile,
      mime_type: delivered.mime_type,
      encoding: delivered.encoding,
      truncated: delivered.truncated || undefined,
    };
    if (includeContent) {
      packaged.content = delivered.content;
      packaged.content_base64 = delivered.content_base64;
    }
    if (includeHashes) {
      packaged.sha256 = await sha256File(workspace, item.path);
    }
    if (delivered.truncated) {
      truncated = true;
    }
    files.push(packaged);
  }

  const summary = includeSummary
    ? await workspace.summarize({
        path: basePath,
        maxFiles,
        previewBytes,
        ignore: params.exclude,
      })
    : undefined;
  const search = includeSearch
    ? await workspace.grep({
        pattern: params.query!,
        path: basePath,
        limit: maxFiles,
        maxBytesPerFile,
        ignore: params.exclude,
      })
    : undefined;

  return {
    object: "local_context_manifest",
    root: workspace.root,
    workspace_name: workspace.name,
    generated_at_unix: Math.floor(Date.now() / 1000),
    base_path: basePath,
    file_count: fileStats.length,
    total_bytes: fileStats.reduce((sum, item) => sum + item.size, 0),
    included_bytes: includedBytes,
    truncated,
    files,
    summary,
    search,
  };
}

export function summarizeLocalContextSensitivity(files: Array<Pick<LocalContextFile, "path" | "sensitivity" | "sensitivity_reason">>): {
  highest: LocalPathSensitivity;
  files: LocalPathSensitivityInfo[];
} {
  const order: Record<LocalPathSensitivity, number> = { normal: 0, sensitive: 1, secret: 2 };
  let highest: LocalPathSensitivity = "normal";
  const sensitiveFiles: LocalPathSensitivityInfo[] = [];
  for (const file of files) {
    if (order[file.sensitivity] > order[highest]) {
      highest = file.sensitivity;
    }
    if (file.sensitivity !== "normal") {
      sensitiveFiles.push({
        path: file.path,
        sensitivity: file.sensitivity,
        reason: file.sensitivity_reason,
      });
    }
  }
  return { highest, files: sensitiveFiles };
}

function matchesPathFilters(relativePath: string, include?: string[], exclude?: string[]): boolean {
  if (include?.length && !include.some((pattern) => pathMatches(relativePath, pattern))) {
    return false;
  }
  if (exclude?.some((pattern) => pathMatches(relativePath, pattern))) {
    return false;
  }
  return true;
}

function pathMatches(relativePath: string, pattern: string): boolean {
  const clean = pattern.replace(/\\/g, "/").replace(/^\/+/, "");
  if (!clean || clean === ".") {
    return true;
  }
  if (!clean.includes("*")) {
    return relativePath === clean || relativePath.startsWith(`${clean}/`);
  }
  const escaped = clean
    .split("*")
    .map((part) => part.replace(/[.+?^${}()|[\]\\]/g, "\\$&"))
    .join(".*");
  return new RegExp(`^${escaped}$`).test(relativePath);
}

function positiveInt(value: number | undefined, fallback: number): number {
  const parsed = Number(value);
  return Number.isFinite(parsed) && parsed > 0 ? Math.trunc(parsed) : fallback;
}

async function sha256File(workspace: LocalWorkspace, relativePath: string): Promise<string> {
  const fullPath = workspace.files.resolvePath(relativePath);
  return createHash("sha256").update(await readFile(fullPath)).digest("hex");
}

export type LocalContextSnapshotFile = LocalWorkspaceSnapshotFile;

File: python/src/agent_api/local/context.py
from __future__ import annotations

import base64
import hashlib
import time
from typing import Any

from .paths import classify_local_path_sensitivity, matches_path_filters, positive_int
from .workdir import LocalWorkdir


def create_local_context_package(workdir: LocalWorkdir, **params: Any) -> dict[str, Any]:
    base_path = params.get("path", ".")
    max_files = positive_int(params.get("max_files"), 80)
    max_bytes = positive_int(params.get("max_bytes"), 256 * 1024)
    max_bytes_per_file = positive_int(params.get("max_bytes_per_file"), 32 * 1024)
    max_depth = positive_int(params.get("max_depth"), 3)
    preview_bytes = positive_int(params.get("preview_bytes"), max_bytes_per_file)
    include_content = params.get("include_content", True)
    include_hashes = params.get("include_hashes", True)
    include_summary = params.get("include_summary", True)
    include_search = bool(params.get("include_search") and str(params.get("query", "")).strip())
    include_secrets = params.get("include_secrets", False)

    scan_stats, scan_warnings = workdir.list_with_warnings(base_path, recursive=True, max_depth=max_depth)
    stats = [
        item
        for item in scan_stats
        if item.type == "file" and matches_path_filters(item.path, params.get("include"), params.get("exclude"))
    ]
    stats.sort(key=lambda item: item.path)
    files: list[dict[str, Any]] = []
    included_bytes = 0
    truncated = len(stats) > max_files

    for item in stats[:max_files]:
        sensitivity = classify_local_path_sensitivity(item.path)
        packaged: dict[str, Any] = {
            "path": item.path,
            "size": item.size,
            "sensitivity": sensitivity["sensitivity"],
            "sensitivity_reason": sensitivity.get("reason"),
        }
        if not include_secrets and sensitivity["sensitivity"] == "secret":
            packaged["omitted_reason"] = "secret_path"
            files.append(packaged)
            continue
        if item.size > max_bytes_per_file:
            packaged["omitted_reason"] = "file_too_large"
            files.append(packaged)
            truncated = True
            continue
        if included_bytes >= max_bytes:
            packaged["omitted_reason"] = "package_budget_exceeded"
            files.append(packaged)
            truncated = True
            continue

        read_budget = min(preview_bytes, max_bytes_per_file, max_bytes - included_bytes)
        if read_budget <= 0:
            packaged["omitted_reason"] = "package_budget_exceeded"
            files.append(packaged)
            truncated = True
            continue

        delivered = workdir.read_file(item.path, max_bytes=read_budget)
        if delivered.get("encoding") == "text":
            included_bytes += len(delivered.get("content", "").encode("utf-8"))
        else:
            included_bytes += len(base64.b64decode(delivered.get("content_base64", "")))

        packaged.update(
            {
                "mime_type": delivered.get("mime_type"),
                "encoding": delivered.get("encoding"),
                "truncated": delivered.get("truncated") or None,
            }
        )
        if include_content:
            if "content" in delivered:
                packaged["content"] = delivered["content"]
            if "content_base64" in delivered:
                packaged["content_base64"] = delivered["content_base64"]
        if include_hashes:
            packaged["sha256"] = hashlib.sha256(workdir.files.resolve_path(item.path).read_bytes()).hexdigest()
        if delivered.get("truncated"):
            truncated = True
        files.append(packaged)

    manifest = {
        "object": "local_context_manifest",
        "root": str(workdir.root),
        "workdir_name": workdir.name,
        "generated_at_unix": int(time.time()),
        "base_path": str(base_path),
        "file_count": len(stats),
        "total_bytes": sum(item.size for item in stats),
        "included_bytes": included_bytes,
        "truncated": truncated,
        "files": files,
        "summary": workdir.summarize(path=base_path, max_files=max_files, max_depth=max_depth, preview_bytes=preview_bytes, ignore=params.get("exclude")) if include_summary else None,
        "search": workdir.grep(pattern=params["query"], path=base_path, limit=max_files, max_bytes_per_file=max_bytes_per_file, ignore=params.get("exclude")) if include_search else None,
    }
    if scan_warnings:
        manifest["scan_warnings"] = scan_warnings
    return manifest
